open-craft/xblocker
# Blockstore-based XBlock runtime

This is a blockstore-based XBlock runtime. It runs as a separate process outside of the LMS. It currently has feature parity with https://github.com/edx/edx-platform/pull/19136

**I'm not planning to develop this any further at the moment**, because converting all XBlocks to Python3, and making them work outside of edx-platform, and sending score/completion/submission events to the LMS asynchronously is too much to take on at once. But let's keep this in mind as a possible future goal :)

Notes:
- Written in Python3, so currently only supports XBlocks that are Python3 compatible. Most XBlocks can be converted quite easily.

**Testing instructions**:
1. [Set up `blockstore` on your docker devstack](https://github.com/open-craft/blockstore#using-with-docker-devstack)
1. Also install and configure [openedx-blockstore-relay](https://github.com/open-craft/openedx-blockstore-relay)
1. Clone this folder into your `devstack` directory.
1. From this repo's folder, run `make dev.up`, `make dev.provision`
1. Run `make xblocker-shell` then `pip install -e /edx/src/whatever-xblock` to install any XBlocks you want to test with (must be Python3 compatible)
1. Create a course in Studio with a Drag and Drop v2 XBlock. Customize the drag and drop XBlock, changing the title, descriptions, mode, and other settings (but don't change the image). Publish the changes.
1. Import the drag and drop xblock into blockstore. See the [relay README](https://github.com/open-craft/openedx-blockstore-relay/blob/master/README.rst) for instructions. The command should look like: `./manage.py cms transfer_to_blockstore --verbosity=2 --collection-uuid "PUT_UUID_HERE" --block-key "block-v1:OpenCraft+B+1+type@drag-and-drop-v2+block@blah"`
1. Note the **UUID** from the resulting bundle. Put them into the URL and visit in in your browser:
    ```
    http://localhost:18222/api/v1/bundles/UUID_GOES_HERE/blocks/
    ```
1. The resulting API response should include a URL for each XBlock in the bundle, like http://localhost:18222/api/v1/blocks/gblock-v1:ec2566b7-e618-429b-8d97-b3ded042b862:drag-and-drop-v2:971aadaaaa5a4886b37fee3ced711b30/ . Go to that URL. Check the resulting HTML / fragment data.
